postgres-protocol/src/types/numeric.rs

```diff
@@ -0,0 +1,1583 @@
+            )?;
+
+            remaining_decimals = 0;
+            break;
+        }
+    }
+
+    if remaining_decimals > 0 {
+        write!(f, "{:0width$}", 0, width = remaining_decimals as usize)?;
+    }
+
+    Ok(())
+}
+
+type ParseError = StdBox<dyn Error + Sync + Send>;
+
+/// Components of a parsed numeric string
+struct ParsedComponents<'a> {
+    sign: NumericSign,
+    integer_part: &'a [u8],
+    decimal_part: &'a [u8],
+    exponent: i32,
+}
+
+impl<'a> ParsedComponents<'a> {
+    /// Calculate the scale for a numeric value, handling scientific notation
+    fn calculate_scale(&self, effective_scale: u16) -> u16 {
+        // Check if the original number is zero
+        let is_zero = self.integer_part.iter().all(|&b| b == b'0')
+            && (self.decimal_part.is_empty() || self.decimal_part.iter().all(|&b| b == b'0'));
+        if is_zero {
+            if self.exponent < 0 {
+                // For "0e-10", preserve the scale implied by the exponent
+                (-self.exponent as u16).max(self.decimal_part.len() as u16)
+            } else {
+                // For "0.0e5", preserve the original decimal scale
+                self.decimal_part.len() as u16
+            }
+        } else if self.exponent < 0 {
+            // For negative exponents, calculate implied decimal places
+            let implied = self.decimal_part.len() as i32 + (-self.exponent);
+            implied.max(0).min(u16::MAX as i32) as u16
+        } else {
+            // For positive exponents, use the effective scale
+            effective_scale
+        }
+    }
+}
+
+fn parse_special_values(s: &str) -> Result<Option<Numeric>, ParseError> {
+    if s.eq_ignore_ascii_case("NaN") {
+        return Ok(Some(Numeric::nan()));
+    }
+    if s.eq_ignore_ascii_case("Infinity") || s.eq_ignore_ascii_case("Inf") {
+        return Ok(Some(Numeric::infinity()));
+    }
+    if s.eq_ignore_ascii_case("-Infinity") || s.eq_ignore_ascii_case("-Inf") {
+        return Ok(Some(Numeric::negative_infinity()));
+    }
+    Ok(None)
+}
+
+fn parse_scientific_component(s: &[u8]) -> Result<(&[u8], Option<i32>), ParseError> {
+    let (s, e) = split_e(s);
+
+    let exp = if let Some(mut e) = e {
+        if e.is_empty() {
+            return Err("empty scientific notation string".into());
+        }
+
+        let mut positive = true;
+        let mut exp: i32 = 0;
+
+        if let Some(&b'-') = e.first() {
+            positive = false;
+            e = &e[1..];
+        } else if let Some(&b'+') = e.first() {
+            e = &e[1..];
+        }
+
+        if e.is_empty() {
+            return Err("empty scientific notation exponent".into());
+        }
+
+        for &b in e {
+            if !b.is_ascii_digit() {
+                return Err("scientific notation string contain non-digit character".into());
+            }
+            // Prevent integer overflow in exponent
+            exp = exp.saturating_mul(10).saturating_add((b - b'0') as i32);
```

Comment:
Can this really overflow if we're checking for `exp` not to get too big in the next line?

Reply:
very true, good catch!